src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+    /// [compaction_frontier, upper)
+    ///
+    /// TODO: the approach to compacting is likely very suboptimal.
+    fn compact(&mut self) -> Result<(), anyhow::Error> {
+        self.consume_wal()?;
+
+        if self.batches.len() > 10 {
+            let batches = std::mem::replace(&mut self.batches, vec![]);
+            let batch = Batch::compact(&batches, &self.trace_path, self.compaction)?;
+            self.batches.push(batch);
+
+            // TODO: This seems like potentially a place with a weird failure mode.
+            for batch in batches {
+                fs::remove_file(&batch.path).with_context(|| {
+                    format!("failed to remove replaced batch {}", batch.path.display())
+                })?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Re-initialize a trace based on the available batch files on disk.
+    pub fn resume(
+        id: GlobalId,
+        traces_path: &Path,
+        wals_path: &Path,
+    ) -> Result<Self, anyhow::Error> {
+        // Need to instantiate a new trace and figure out what batches
+        // we have access to.
+        let trace_path = traces_path.join(id.to_string());
+        let wal_path = wals_path.join(id.to_string());
+        let mut ret = Self {
+            trace_path,
+            wal_path,
+            batches: Vec::new(),
+            compaction: None,
+        };
+
+        let mut batches = ret.find_batches()?;
+        ret.batches.append(&mut batches);
+
+        Ok(ret)
+    }
+
+    /// Read in the data for this relation, from all available batches and
+    /// WAL segments back into memory.
+    pub fn read(&self) -> Result<Vec<Message>, anyhow::Error> {
+        let mut out = vec![];
+
+        for batch in self.batches.iter() {
+            let mut messages = batch.read()?;
+            out.append(&mut messages);
+        }
+
+        let finished_segments = self.find_finished_wal_segments()?;
+        let unfinished_segment = self.find_unfinished_wal_segment()?;
+
+        for segment in finished_segments {
+            let mut messages = read_segment(&segment)?;
+            out.append(&mut messages);
+        }
+
+        out.append(&mut read_segment(&unfinished_segment)?);
+
+        out.dedup();
+        Ok(out)
+    }
+}
+
+pub struct Compacter {
+    rx: mpsc::UnboundedReceiver<CompacterMessage>,
+    traces: HashMap<GlobalId, Trace>,
+    traces_path: PathBuf,
+    wals_path: PathBuf,
+}
+
+impl Compacter {
+    pub fn new(
+        rx: mpsc::UnboundedReceiver<CompacterMessage>,
+        traces_path: PathBuf,
+        wals_path: PathBuf,
+    ) -> Result<Self, anyhow::Error> {
+        Ok(Self {
+            rx,
+            traces: HashMap::new(),
+            traces_path,
+            wals_path,
+        })
+    }
+
+    async fn compact(&mut self) -> Result<(), anyhow::Error> {
+        let mut interval = tokio::time::interval(COMPACTER_INTERVAL);
+        loop {
+            select! {
+                data = self.rx.recv() => {
```

Comment:
I don't understand this construction. It looks like some conditional call that binds a thing to `data`. But it also looks like it blocks on `recv()` rather than awaiting it. What's going on in here?

Reply:
so the select! macro is awaiting both recv-ing the data and the interval concurrently iirc -- tbh I cribbed this construction from elsewhere I looked at this doc https://docs.rs/tokio/1.2.0/tokio/macro.select.html this seems to be a ergonomic-ish way to await multiple things at once from the same tokio task?